elasticdl/python/common/save_utils.py

```diff
@@ -0,0 +1,256 @@
+import os
+import tempfile
+
+import tensorflow as tf
+
+from elasticdl.proto import elasticdl_pb2
+from elasticdl.python.common.hash_utils import int_to_id, string_to_id
+from elasticdl.python.common.tensor import Tensor
+from elasticdl.python.ps.embedding_table import create_embedding_table
+from elasticdl.python.ps.parameters import Parameters
+
+
+def save_pb_to_file(pb_obj, file_name):
+    """Save a protobuf object to file"""
+    encoded_model = pb_obj.SerializeToString()
+    with open(file_name, "wb") as f:
+        f.write(encoded_model)
+
+
+def load_pb_from_file(pb_obj, file_name):
+    """Load a protobuf object from a file"""
+    with open(file_name, "rb") as f:
+        pb_obj.ParseFromString(f.read())
+    return pb_obj
+
+
+def _get_params_shard_from_pb(model_pb, shard_index, shard_num):
+    """Get parameters including variables values and embedding table
+    from a model protobuf.
+
+    Args:
+        model_pb: A Model protobuf instance.
+        shard_index: Model shard index.
+        shard_num: The total number of model shards.
+
+    Return:
+        non_embedding_vars: A Python dict in which the key is a variable
+            name and the value is a `tf.Variable` object.
+        embedding_table_values: A Python dict in which the key is an embedding
+            table name and the value is a tuple with 2 elements. The value[0]
+            is indices and value[1] is the corresponding embedding vector.
+    """
+    non_embedding_vars = {}
+    embedding_table_values = {}
+
+    for tensor_pb in model_pb.param:
+        tensor = Tensor.from_tensor_pb(tensor_pb)
+        if tensor.indices is not None:
+            embedding_table_values.setdefault(tensor.name, ([], []))
+            for embedding_id, vector in zip(tensor.indices, tensor.values):
+                if int_to_id(embedding_id, shard_num) == shard_index:
+                    embedding_table_values[tensor.name][0].append(embedding_id)
+                    embedding_table_values[tensor.name][1].append(vector)
+        else:
+            if string_to_id(tensor.name, shard_num) == shard_index:
+                non_embedding_vars[tensor.name] = tf.Variable(
+                    initial_value=tensor.values, trainable=True
+                )
+    return non_embedding_vars, embedding_table_values
+
+
+class Checkpoint(object):
+    def __init__(self, version, file):
+        self.version = version
+        self.file = file
+
+
+class CheckpointSaver(object):
```

Comment:
Can this implementation be generic instead of specific for multi-ps?

Reply:
The `CheckpointSaver` is not specific for multi-ps. It takes charge of saving the parameters including embedding and non-embedding variables using `proto.Model` in the current process. For all-reduce and other strategies, we can also use the implementation to save checkpoints. Using all-reduce strategy, any worker can create a `proto.Model` with its all parameters and call `CheckpointSaver.save` with `shard_index=0` and `shard_num=1`. We also can restore a `Parameters` instance from a checkpoint file and assign those parameters to the model in workers.